### PR TITLE
kopt: correctly handle CJK character detection for space insertion

### LINKED ISSUE #8433: No word spacing when adding a note on a PDF
* KOReader version: latest nightly (59)
* Device: kobo libra 2 / android

#### Issue
hi, i use nightly koreader on my devices, today i face a weird issue. When im highlighting a text to add a note in it, koreader seems cant recognize the spacing between text. I tried to emulate this issue on a stable build (10.1), but it seems stable build didnt have this issue. Ive attached screenshots from each stable and nightly builds to explain better.

<a href="https://ibb.co/6WrZKCd">stable-build-10-1</a>
Stable build (10.1)

<a href="https://ibb.co/5TvxkWv">nightly-build-59</a>
Nightly build (59)

##############################################
ive also attached sample pages from pdf i use on this issue
[test page.pdf](https://github.com/koreader/koreader/files/7506898/test.page.pdf)

Cheers!

### PR DESCRIPTION
**Pull request by cyphar**
Previously getTextFromBoxes would just pass the first and last three
bytes of the current and previous words when trying to detect CJK
characters (which shouldn't have spaces inserted).

However, this handling was not correct because CJK characters can be
longer than 3 bytes, and internally BaseUtil.utf8charcode doesn't ensure
that it was only given a single utf8 character (it blindly does the bit
operations on whatever length code you give it).

As a result, before this patch selections in PDF documents would have
lots of spaces stripped because getTextFromBoxes would think that almost
all characters were CJK characters.

Fixes: 6f1b70e5eb71 ("util.utf8: improve CJK character detection")
Fixes #8433 
Signed-off-by: Aleksa Sarai <cyphar@cyphar.com>

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/8438)

### CODE REVIEW

**Comment by cyphar** [comment]
This is the only other user of `isCJKChar` so there shouldn't be any other cases of this lurking around. However, I think `utf8charcode` should probably be hardened to actually check that the length of the passed "character" actually matches the character that was passed.

File: frontend/document/koptinterface.lua [hunk 1]
```diff
@@ -14,6 +14,7 @@ local KOPTContext = require("ffi/koptcontext")
 local Persist = require("persist")
 local TileCacheItem = require("document/tilecacheitem")
 local logger = require("logger")
+local textutil = require("util")
 local util = require("ffi/util")
```
Comment (poire-z) [inline frontend/document/koptinterface.lua:18]: `textutil` would be a newcomer for naming this old module :)
Best to just use like in many other modules:
```lua
local util = require("util")
local FFIUtil or BaseUtil = require("ffi/util") -- we use any of these 2 names in other modules
```
The only usage in koptinterface of FFIUtil is util.usleep, some commented out util.gettime, and util.isAndroid - so you're allowed to update these 4 unrelated other lines for consistency sake :)

Comment (cyphar) [reply frontend/document/koptinterface.lua:18]: Okay, will do. :smile:

**Review by poire-z — APPROVED** [review]

File: frontend/document/koptinterface.lua [hunk 2]
```diff
@@ -968,10 +968,11 @@ function KoptInterface:getTextFromBoxes(boxes, pos0, pos1)
...
                         add_space = false
                     elseif dist_from_prev_word < box_height * 0.8 then
-                        if isCJKChar(prev_word:sub(-3, -1)) and isCJKChar(word:sub(1, 3)) then
-                            -- Two CJK chars whose spacing is not large enough
-                            -- (we checked the 3 UTF8 bytes that CJK chars must be,
-                            -- no need to split into unicode codepoints)
+                        local prev_word_end = prev_word:match(util.UTF8_CHAR_PATTERN.."$")
+                        local word_start = word:match(util.UTF8_CHAR_PATTERN)
+                        if util.isCJKChar(prev_word_end) and util.isCJKChar(word_start) then
+                            -- Two CJK chars whose spacing is not large enough,
```
Comment (Frenzie) [inline frontend/document/koptinterface.lua:974]: Anthropomorphizing the CJK characters? :-P

Comment (poire-z) [reply frontend/document/koptinterface.lua:974]: Time for my English lesson :) How would you write this ?
I guess it's about `whose` ?
https://www.merriam-webster.com/words-at-play/whose-used-for-inanimate-objects
https://english.stackexchange.com/questions/23541/can-whose-refer-to-an-inanimate-object

(But in my childish heart, I consider CJK chars animals, like octopuses and ants, whose legs are many :)

Comment (cyphar) [reply frontend/document/koptinterface.lua:974]: The other option would be "Two CJK chars which don't have a large enough spacing" or something like that. To me both sound okay, but "whose" is more concise which is why I wrote it that way :smile:. While "whose" does somewhat anthropomorphise them it didn't/doesn't sound particularly off to me.
